File: ai-service/secrets_loader.py
import os
import sys
import uvicorn
from azure.identity import DefaultAzureCredential
from azure.keyvault.secrets import SecretClient

def load_secrets():
    kv_name = os.environ.get("KEY_VAULT_NAME")
    if kv_name:
        print(f"Loading secrets from Key Vault: {kv_name} using Azure Workload Identity...")
        kv_uri = f"https://{kv_name}.vault.azure.net"
        credential = DefaultAzureCredential()
        client = SecretClient(vault_url=kv_uri, credential=credential)
        try:
            secrets = client.list_properties_of_secrets()
            for secret_prop in secrets:
                secret_name = secret_prop.name
                secret = client.get_secret(secret_name)
                env_name = secret_name.upper().replace("-", "_")
                os.environ[env_name] = secret.value
                
                # Custom alias mapping
                if env_name == "AI_PORT":
                    os.environ["PORT"] = secret.value
            print("Successfully loaded secrets from Key Vault.")
        except Exception as e:
            print(f"Error loading secrets from Key Vault: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        print("KEY_VAULT_NAME not set. Sourcing from environment directly.")

if __name__ == "__main__":
    load_secrets()
    # Now run uvicorn
    port = int(os.environ.get("PORT", 8000))
    uvicorn.run("main:app", host="0.0.0.0", port=port)

File: exam-service/src/secrets-loader.ts
import { DefaultAzureCredential } from '@azure/identity';
import { SecretClient } from '@azure/keyvault-secrets';

async function bootstrap() {
  const keyVaultName = process.env.KEY_VAULT_NAME;
  if (keyVaultName) {
    console.log(`Loading secrets from Key Vault: ${keyVaultName} using Azure Workload Identity...`);
    const credential = new DefaultAzureCredential();
    const client = new SecretClient(`https://${keyVaultName}.vault.azure.net`, credential);
    try {
      for await (const secretProperties of client.listPropertiesOfSecrets()) {
        const secretName = secretProperties.name;
        const secret = await client.getSecret(secretName);
        const envName = secretName.toUpperCase().replace(/-/g, '_');
        process.env[envName] = secret.value;
        
        // Map common aliases
        if (envName === 'DB_CONNECTION_STRING' || envName === 'DATABASE_URL') {
          process.env.DATABASE_URL = secret.value;
        }
        if (envName === 'REDIS_CONNECTION_STRING' || envName === 'REDIS_URL') {
          process.env.REDIS_URL = secret.value;
        }
        if (envName === 'JWT_ACCESS_SECRET' || envName === 'JWT_SECRET') {
          process.env.JWT_SECRET = secret.value;
          process.env.JWT_ACCESS_SECRET = secret.value;
        }
      }
      console.log('Successfully loaded all secrets from Key Vault.');
    } catch (error) {
      console.error('Error loading secrets from Key Vault:', error);
      process.exit(1);
    }
  } else {
    console.log('KEY_VAULT_NAME not set. Sourcing secrets from local environment.');
  }
}

bootstrap().then(() => {
  require('./index.js');
}).catch((err) => {
  console.error('Unhandled error in secrets loader:', err);
  process.exit(1);
});
